### PR TITLE
lig-3245: Introduce new generated code

### PR DESCRIPTION
**Pull request by huan-lightly-0**
Python API client generated with `openapi-generator`. This new code generator provides better support for validation and type hinting.

### CODE REVIEW

File: lightly/openapi_generated/swagger_client/models/access_role.py
Comment (guarin) [inline lightly/openapi_generated/swagger_client/models/access_role.py:1]: Why is this file removed? Is the class now stored somewhere else?

Comment (huan-lightly-0) [reply lightly/openapi_generated/swagger_client/models/access_role.py:1]: It's not stored somewhere else but simply removed. The reason is that this "class" is actually just a string. It still has validation, for example, in [DatasetData](https://github.com/lightly-ai/lightly/pull/1261/files/6f495ae62ee473a4ba1b55832057f5b3c3fc6c7a#diff-fda8a22c8bbf1cb3fa5abd9a98dfc783f8df666dc2d6c73cdfcc77b8d8680113R74).
Classes that are actually native types fall back to native types.

File: lightly/api/api_workflow_artifacts.py [hunk 1]
```diff
@@ -1,12 +1,10 @@
 import os
 
 from lightly.api import download
-from lightly.openapi_generated.swagger_client import (
+from lightly.openapi_generated.swagger_client.models import (
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/api/api_workflow_artifacts.py:4]: Edit: Looking at the `__init__` file in swagger client the old imports should still work.

[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply lightly/api/api_workflow_artifacts.py:4]: It is possible, but I think it's better to reference things from submodules. It looks clearer to me. This is not necessary though.

File: lightly/api/swagger_rest_client.py [hunk 2]
```diff
@@ -53,20 +53,18 @@ def request(
...
-        # Set default timeout. This is necessary because the swagger api client does not
+        # Set default timeout. This is necessary because the openapi client does not
         # respect timeouts configured by urllib3. Instead it expects a timeout to be
         # passed with every request. See code here:
         # https://github.com/lightly-ai/lightly/blob/ffbd32fe82f76b37c8ac497640355314474bfc3b/lightly/openapi_generated/swagger_client/rest.py#L141-L148
         if _request_timeout is None:
             _request_timeout = self.timeout
 
-        flat_query_params = _flatten_list_query_parameters(query_params=query_params)
-
```
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply lightly/api/swagger_rest_client.py:64]: Yes. The generated client already does something similar to `_flatten_list_query_parameters`. I've tested this.

File: lightly/api/api_workflow_compute_worker.py [hunk 3]
```diff
@@ -666,14 +666,9 @@ def _validate_config(
...
 
-    if not hasattr(type(obj), "swagger_types"):
-        raise TypeError(
-            f"Type {type(obj)} of argument 'obj' has not attribute 'swagger_types'"
-        )
-
     for key, item in cfg.items():
         if not hasattr(obj, key):
-            possible_options = list(type(obj).swagger_types.keys())
+            possible_options = list(type(obj).__fields__.keys())
```
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply lightly/api/api_workflow_compute_worker.py]: It is. Updated.

File: tests/api_workflow/test_api_workflow_tags.py [hunk 4]
```diff
@@ -1,89 +1,214 @@
...
+    mocked_api = mocker.MagicMock()
+
+    client = ApiWorkflowClient()
+    client._tags_api = mocked_api
+    client._dataset_id = dataset_id
+    client._creator = TagCreator.UNKNOWN
+    client.create_tag_from_filenames(fnames_new_tag=["file2"], new_tag_name="some-tag")
+    mocked_get_filenames.assert_called_once()
+    mocked_api.create_tag_by_dataset_id.assert_called_once()
+    kwargs = mocked_api.create_tag_by_dataset_id.call_args[1]
```
Comment (guarin) [inline tests/api_workflow/test_api_workflow_tags.py:46]: FYI, you could use `kwargs = mocked_api.create_tag_by_dataset_id.call_args.kwargs` instead.

File: tests/api_workflow/test_api_workflow_upload_dataset.py [hunk 5]
```diff
@@ -1,22 +1,23 @@
...
-    SamplePartialMode,
-)
+from lightly.openapi_generated.swagger_client.models import SamplePartialMode
 from tests.api_workflow.mocked_api_workflow_client import MockedApiWorkflowSetup
 
 
+# TODO: fix this text
+@pytest.skip(
+    "Skip this test for now. Test cases need to be updated.", allow_module_level=True
+)
```
Comment (guarin) [inline tests/api_workflow/test_api_workflow_upload_dataset.py:20]: I believe the test and corresponding code can safely be deleted.

File: tests/api_workflow/utils.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import random
+
+# [a-f0-9]
+_CHARACTER_SET = [str(i) for i in range(10)] + [chr(i + 97) for i in range(6)]
```
Comment (guarin) [inline tests/api_workflow/utils.py:4]:
```suggestion
_CHARACTER_SET = "abcdef0123456789"
```
Might be a bit easier to understand :)

File: tests/cli/test_cli_download.py [hunk 7]
```diff
@@ -13,6 +13,7 @@
 )
 
 
+@pytest.skip("Skip this test.", allow_module_level=True)
```
Comment (guarin) [inline tests/cli/test_cli_download.py:16]: Why do we have to skip this test?

File: tests/cli/test_cli_magic.py [hunk 8]
```diff
@@ -14,6 +15,7 @@
 )
 
 
+@pytest.skip("Skip this test.", allow_module_level=True)
```
Comment (guarin) [inline tests/cli/test_cli_magic.py:18]: Why do we have to skip this test?

**Review by guarin — COMMENTED** [review]
Went through the tests as well :) It is a bit hard to review as there are many changes but thanks a lot for converting many of them to pytest format!

File: tests/api_workflow/test_api_workflow_compute_worker.py [hunk 9]
```diff
@@ -145,7 +150,7 @@ def test_schedule_compute_worker_run__priority(self):
 
     def test_schedule_compute_worker_run__runs_on(self):
         scheduled_run_id = self.api_workflow_client.schedule_compute_worker_run(
-            worker_config={}, lightly_config={}, runs_on=["AAA", "BBB"]
+            worker_config={}, lightly_config={}, runs_on=[generate_id(), generate_id()]
```
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_compute_worker.py:153]: This is not necessary. Reverted.

File: tests/api_workflow/test_api_workflow_compute_worker.py [hunk 10]
```diff
@@ -185,7 +202,7 @@ def _check_if_openapi_generated_obj_is_valid(self, obj) -> Any:
 
         return obj_api
 
-    def test_selection_config(self):
+    def xtest_selection_config(self):
```
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_compute_worker.py:205]: Because I'm not really sure if this is necessary. It seems that this tests the generated code and I don't think it makes sense. Can we remove it?

File: tests/api_workflow/test_api_workflow_datasets.py [hunk 11]
```diff
@@ -118,72 +129,206 @@ def test_get_all_datasets(self):
...
+
+def test_delete_dataset(mocker: MockerFixture) -> None:
+    mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
+    mock_datasets_api = mocker.MagicMock()
+    client = ApiWorkflowClient()
+    client._dataset_id = "foo"
+    client._datasets_api = mock_datasets_api
+    client.delete_dataset_by_id("foobar")
+    mock_datasets_api.delete_dataset_by_id.assert_called_once_with(dataset_id="foobar")
+    assert not hasattr(client, "_dataset_id")
```
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_datasets.py:225]: Yes. https://github.com/lightly-ai/lightly/blob/master/lightly/api/api_workflow_datasets.py#L421

File: tests/api_workflow/test_api_workflow_datasets.py [hunk 12]
```diff
@@ -118,72 +129,206 @@ def test_get_all_datasets(self):
...
 
-    def test_create_dataset_with_counter_nonexisting(self):
-        self.api_workflow_client.create_dataset(dataset_name="basename")
-        self.api_workflow_client.create_new_dataset_with_unique_name(
-            dataset_basename="baseName"
-        )
-        assert self.api_workflow_client._datasets_api.datasets[-1].name == "baseName"
+def test_create_new_dataset_with_unique_name__new_name(mocker: MockerFixture) -> None:
+    mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
+    mocker.patch.object(ApiWorkflowClient, "dataset_name_exists", return_value=False)
```
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_datasets.py:135]: > Also it avoids issues with accidentally mocking methods that do not exist on the client.
pytest raises an error when an undefined method is patched if I recall correctly.

I didn't create one helper function for the API client because it requires customisation in different test cases and thus a helper function doesn't seem to save much effort.

File: tests/api_workflow/test_api_workflow_compute_worker.py [hunk 13]
```diff
@@ -282,14 +300,18 @@ def test_selection_config_from_dict() -> None:
...
+        match=r"strategies\n  ensure this value has at least 1 items",
+    ):
+        api_workflow_compute_worker.selection_config_from_dict(cfg)
 
 
 def test_selection_config_from_dict__extra_key() -> None:
     cfg = {"strategies": [], "invalid-key": 0}
     with pytest.raises(
-        TypeError, match="got an unexpected keyword argument 'invalid-key'"
+        ValidationError,
```
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_compute_worker.py:313]: I think it's okay to keep it as is since probably no one catches this `TypeError`. There is little one can do after catching this error.

A separate topic: handling validation errors properly. Ideally, `ValidationError` should be handled by us and reraised because it's from pydantic and sometimes the error message is purely internal.

File: tests/api_workflow/test_api_workflow_upload_dataset.py (hunk 5 above)
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_upload_dataset.py:20]: Will do that cleanup in a separate PR.

File: tests/cli/test_cli_magic.py (hunk 8 above)
Comment (huan-lightly-0) [reply tests/cli/test_cli_magic.py:18]: Because I didn't want to fix the mocked API client. I've fixed that and re-enabled this anyway.

File: tests/cli/test_cli_download.py (hunk 7 above)
Comment (huan-lightly-0) [reply tests/cli/test_cli_download.py:16]: Because I didn't want to fix the mocked API client. I've fixed that and re-enabled this anyway.

File: tests/api_workflow/test_api_workflow_artifacts.py [hunk 14]
```diff
@@ -171,22 +177,41 @@ def test__get_compute_worker_run_checkpoint_url(
...
+        docker_version="",
+        state=DockerRunState.COMPUTING_METADATA,
+        created_at=0,
+        last_modified_at=0,
+        artifacts=[
+            DockerRunArtifactData(
+                id=generate_id(),
+                file_name="report.pdf",
+                type=DockerRunArtifactType.REPORT_PDF,
+            ),
```
[the remark this answers is not in the document]
Comment (guarin) [reply tests/api_workflow/test_api_workflow_artifacts.py:204]: I believe you didn't change it.

File: tests/api_workflow/test_api_workflow_compute_worker.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (guarin) [reply tests/api_workflow/test_api_workflow_compute_worker.py:205]: I believe it is there because we had issues with configs not being properly serialized/deseralized. If this is not an issue with the new generator we can remove the test.

File: tests/api_workflow/test_api_workflow_compute_worker.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (guarin) [reply tests/api_workflow/test_api_workflow_compute_worker.py:313]: Is there a good way to handle the pydantic errors? Because ideally we can show the users exactly where the error happened.

File: tests/api_workflow/test_api_workflow_artifacts.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_artifacts.py]: Sorry. Updated

File: tests/api_workflow/test_api_workflow_datasets.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (guarin) [reply tests/api_workflow/test_api_workflow_datasets.py:135]: Let's leave it like this for now. The benefit of the helper would be that the client is properly initialized and actually has all the instance attributes it normally has. Plus you wouldn't have to patch the full api classes but only their respective methods. So instead of this:
```
mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
client = ApiWorkflowClient()
client._datasets_api = mocker.MagicMock()
```
You could patch only the requested method:
```
mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
client = ApiWorkflowClient()
client._datasets_api.method = mocker.MagicMock()
```

Because with the first method you might assert on methods that don't exist because the whole api is mocked:
```
client._datasets_api.some_other_method.assert_called_with(...)
```

File: tests/api_workflow/test_api_workflow_datasets.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (guarin) [reply tests/api_workflow/test_api_workflow_datasets.py:225]: Oh, I see

File: tests/cli/test_cli_download.py (hunk 7 above)
Comment (guarin) [reply tests/cli/test_cli_download.py]: Awesome thanks!

File: tests/api_workflow/test_api_workflow_compute_worker.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_compute_worker.py:313]: I mean normally it's okay to just leave pydantic errors as is. It's a problem only when users try to catch the pydantic error types because that means they need to import pydantic in their code as well. However, I don't think that's a big issue because no one probably catches this kind of error.
The bigger issue is about errors irrelevant to user input. Users won't understand what went wrong unless they look into the source code. This can be prevented by proper tests, so let's not be too concerned about this now.

File: tests/api_workflow/test_api_workflow_compute_worker.py [hunk 15]
```diff
@@ -520,28 +522,30 @@ def get_compute_worker_run_info(self, scheduled_run_id: str):
 
 
 def test_get_compute_worker_runs(mocker: MockerFixture) -> None:
+    dataset_id = generate_id()
+    run_ids = [generate_id(), generate_id()]
     client = ApiWorkflowClient(token="123")
```
Comment (guarin) [inline tests/api_workflow/test_api_workflow_compute_worker.py:527]: I just noticed this, doesn't this make an api request to prod if we instantiate `ApiWorkflowClient` without mocking it? Because `ApiWorkflowClient` runs a version check on initialization.

File: tests/api_workflow/test_api_workflow_compute_worker.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_compute_worker.py:205]: I kept the test and changed this to `DockerWorkerConfigV3`. It's not an issue.

File: tests/api_workflow/test_api_workflow_compute_worker.py (hunk 15 above)
Comment (huan-lightly-0) [reply tests/api_workflow/test_api_workflow_compute_worker.py:528]: You are right. I just updated these test cases.

There are actually some other worse things:
* Every time `from lightly.... import ...` is executed in the test suites, a version check to prod is performed.
* The `MockedApiWorkflowClient` class never really mocked the version API, so these test cases have been hitting prod for a long time.

Comment (guarin) [reply tests/api_workflow/test_api_workflow_compute_worker.py:528]: Thanks for investigating! Let's make a follow-up and fix asap.

Comment (guarin) [reply tests/api_workflow/test_api_workflow_compute_worker.py:528]: Fix is here #1267

**Review by guarin — APPROVED** [review]
LGTM!
